integration/image/delta_test.go

```diff
@@ -33,16 +36,84 @@ func TestDelta(t *testing.T) {
 		types.ImageDeltaOptions{
 			Tag: delta,
 		})
-	if err != nil {
-		t.Fatalf("Creating delta: %s", err)
-	}
+	assert.NilError(t, err)
 	io.Copy(ioutil.Discard, rc)
 	rc.Close()
 
-	_, _, err = client.ImageInspectWithRaw(ctx, delta)
-	if err != nil {
-		t.Fatalf("Inspecting delta: %s", err)
-	}
+	inspectDelta, _, err := client.ImageInspectWithRaw(ctx, delta)
+	assert.NilError(t, err)
+
+	inspectBase, _, err := client.ImageInspectWithRaw(ctx, base)
+	assert.NilError(t, err)
+	assert.Assert(t, inspectDelta.Config.Labels["io.resin.delta.base"] == inspectBase.ID)
+}
+
+// TestDeltaCreateDestinationLock triggers a delta generation job, waits for it
+// to start and removes the destination image.
+//
+// Until 1e83de47e32ac0caf3b4e02094aeb76da28b90b3 this would remove unprocessed
+// destination layers and break delta generation.
+func TestDeltaCreateDestinationLock(t *testing.T) {
+	var (
+		base   = "debian:8"
+		target = "debian:10"
+		delta  = "debian:delta"
+	)
+
+	var (
+		err    error
+		rc     io.ReadCloser
+		ctx    = context.Background()
+		client = testEnv.APIClient()
+	)
+
+	pullBaseAndTargetImages(t, client, base, target)
+
+	rc, err = client.ImageDelta(ctx,
+		base,
+		target,
+		types.ImageDeltaOptions{
+			Tag: delta,
+		})
+	assert.NilError(t, err)
+	defer rc.Close()
+
+	var (
+		waitFingerprinting = make(chan struct{}, 100)
+		waitDelta          = make(chan struct{})
+	)
+	go func() {
+		defer close(waitFingerprinting)
+		defer close(waitDelta)
+		for br := bufio.NewReader(rc); ; {
+			line, err := br.ReadString('\n')
+			if err != nil {
+				if err == io.EOF {
+					break
+				}
+				t.Fatal(err)
+			}
+			if strings.Contains(line, "Fingerprinting") {
+				waitFingerprinting <- struct{}{}
+			}
+			if strings.Contains(line, "layer does not exist") {
+				t.Fail()
+			}
+		}
+	}()
+
+	<-waitFingerprinting
+	deleted, err := client.ImageRemove(ctx, target, types.ImageRemoveOptions{})
+	assert.NilError(t, err)
+	assert.Assert(t, len(deleted) != 0)
```

Comment:
Hmm, maybe it's just my Docker ignorance, but I expected `len(deleted) == 0` here. My reasoning is the following: we just waited for fingerprinting to start, then immediately tried to delete the target image (while the fingerprinting is supposedly still running). But this target image is locked, so it should not be really deleted. Therefore, I expected `err == nil` (no real errors while running the remove image operation) and `len(deleted) == 0` (no image was removed).

And this expectation of mine is not correct, what does `len(deleted)` really means? In which way this new assertion increases our confidence that image locking during delta is working as intended?

Reply:
no you are 100% correct. it should *not* remove anything here, as it says in the comment to the test function

Reply:
The code looks right to me now, but when I run the test on my machine, the `assert.Assert(t, len(deleted) == 0)` is failing. Does it work for you?